fixed struct contract violation message

tommymchugh opened the pull request:
Resolves https://github.com/racket/racket/issues/2998

Example:
```
#lang racket
(define-struct/contract s ([x integer?][y boolean?]))
(s 1 2)
```

Old Output:
```
$  racket ~/tmp.rkt
make-s: contract violation
  expected: boolean?
  given: 2
  in: the 2nd argument of
      (-> integer? boolean? symbol? any)
  contract from: (struct s)
  blaming: /Users/tommymchugh/test.rkt
   (assuming the contract is correct)
  at: /Users/tommymchugh/test.rkt
```

New Output:
```
$  racket ~/tmp.rkt
make-s: contract violation
  expected: boolean?
  given: 2
  in: the 2nd argument of
      (-> integer? boolean? any)
  contract from: make-s
  blaming: /Users/tommymchugh/test.rkt
   (assuming the contract is correct)
  at: /Users/tommymchugh/test.rkt
```

AlexKnauth commented on racket/collects/racket/contract/region.rkt:
```diff
@@ -477,21 +477,24 @@
                                             (begin
                                               (define-values () (begin auto-check ... (values)))
-                                              (define (guard super-field ... non-auto-name ... struct-name)
+                                              (define (guard-contract super-field ... non-auto-name ...)
```
Is there now any difference between this `guard-contract` and `values`?

The only differences I'm aware of are arity and name. The arity shouldn't matter here. The name shouldn't matter either, as the error messages come from the `contract` form which supplies the name with `(quote maker)`.

So you should be able to get by without defining `guard-contract`, with
```
(define (guard super-field ... non-auto-name ... struct-name)
  ((contract (-> super-contract ... non-auto-contracts ... any)
             values
             ....)
   super-field ... non-auto-name ...))
```

One more minor question, I noticed you replaced `(current-contract-region)` with `(quote maker)`. I don't think there's good reason to change it. Since it's a syntax parameter not a run-time parameter, it won't be affected by being inside the function body.

tommymchugh replied:
Yes, made these changes, but this will probably change anyway with the new `->`

rfindler commented:
It isn't a good idea to put the call to `contract` inside the guard; that does too much work to put done on each call. 

I think the right approach here is to make (yet another) variation of `->` that ignores the last argument when it produces the error message.

rfindler commented:
(also, of course, we need test cases!)

tommymchugh commented:
@rfindler Roger that, will work on that and update this PR.

AlexKnauth commented:
> It isn't a good idea to put the call to `contract` inside the guard

So is it better to put the `contract` call outside, even when the guard calls that result function every time like this:
```
(define guarded-values
  (contract (-> super-contract ... non-auto-contracts ... any)
            values
            ....))
(define (guard super-field ... non-auto-name ... struct-name)
  (guarded-values super-field ... non-auto-name ...))
```
Is this okay, or is there a downside that I'm not seeing, which would make a new variation of `->` better?

rfindler commented:
@AlexKnauth Oh, of course. That's better! Thanks.

tommymchugh commented:
That sounds great, just pushed those changes to this PR.

AlexKnauth commented:
I'm still wondering about the blame information.

The old `positive-blame-expr` was the `(current-contract-region)` within the `with-contract #:region struct struct-name` form, which in the old error message example was `(struct s)`.
```
  contract from: (struct s)
```

However, the new `positive-blame-expr` is now `(current-contract-region)` outside of that with-contract, so in the new error message example it's more generic:
```
  contract from: /path/to/test.rkt
```

So to keep the positive blame party the same as it was before, the `contract` expression for `guarded-values` should be kept within the `with-contract #:region struct struct-name` form.

tommymchugh commented:
@AlexKnauth That did the trick! Thanks so much for the advice. That was what I had noticed when I had initially changed the blame-id since it was outside the struct, but now there is no regression when the `define` is within the `with-contract`. Violation message is now exactly like previous version, just without the extra argument in the procedure. I just squashed and rebased, so looks pretty much good to go unless there are any other concerns I can address!

rfindler commented:
There are no new test cases and the new behavior certainly can be captured with tests.

tommymchugh commented:
Sounds good, will add those!

On Wed, Jan 8, 2020 at 7:33 AM Robby Findler <notifications@github.com>
wrote:

> There are no new test cases and the new behavior certainly can be captured
> with tests.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/racket/racket/pull/2999?email_source=notifications&email_token=AAMRJXNW2VPMCZ2FSIGKSMLQ4XI25A5CNFSM4KC46BT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIMNHJI#issuecomment-572052389>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMRJXPU6GOARQS4TJZVWILQ4XI25ANCNFSM4KC46BTQ>
> .
>
-- 
*Thomas McHugh*
*Northwestern University*
Undergraduate Student
Email: mchugh@u.northwestern.edu
Phone: 847-922-4801

rfindler commented:
(Also we'll want a test case that captures the issue that @AlexKnauth noticed, too.)

pmatos commented:
ping! Can we look at this again so it is not forgotten? There's an approved review by @AlexKnauth but no merge.

rfindler commented:
Looks to me like it isn't ready to merge based on the discussion, @pmatos .

tommymchugh commented:
@pmatos @rfindler Sorry kinda lost track of this with all the craziness of school and such! I'll work on this and submit a patch ASAP! :)

rfindler commented:
No rush! Thanks for your contributions.
